File: src/research_artifact_uploader/gitinfo.py
from __future__ import annotations

import subprocess
from dataclasses import dataclass
from pathlib import Path


@dataclass
class GitInfo:
    available: bool
    repo_root: str | None = None
    commit: str | None = None
    branch: str | None = None
    dirty: bool | None = None


def _run_git(cmd: list[str], cwd: Path | None = None) -> str:
    return subprocess.check_output(
        cmd,
        cwd=str(cwd) if cwd is not None else None,
        text=True,
        stderr=subprocess.DEVNULL,
    ).strip()


def read_git_info(cwd: Path | None = None) -> GitInfo:
    try:
        return GitInfo(
            available=True,
            repo_root=_run_git(["git", "rev-parse", "--show-toplevel"], cwd=cwd),
            commit=_run_git(["git", "rev-parse", "HEAD"], cwd=cwd),
            branch=_run_git(["git", "branch", "--show-current"], cwd=cwd),
            dirty=bool(_run_git(["git", "status", "--porcelain"], cwd=cwd)),
        )
    except Exception:
        return GitInfo(available=False)

File: src/gitinfo.rs
use std::path::Path;
use std::process::Command;

use crate::AppResult;

#[derive(Debug)]
pub struct GitInfo {
    pub available: bool,
    pub repo_root: Option<String>,
    pub commit: Option<String>,
    pub branch: Option<String>,
    pub dirty: Option<bool>,
}

pub fn read_git_info() -> GitInfo {
    let root = run_git(["rev-parse", "--show-toplevel"].as_slice())
        .and_then(|s| Ok(s.trim().to_string()))
        .ok();
    if root.is_none() {
        return GitInfo {
            available: false,
            repo_root: None,
            commit: None,
            branch: None,
            dirty: None,
        };
    }

    let commit = run_git(["rev-parse", "HEAD"].as_slice()).ok();
    let branch = run_git(["branch", "--show-current"].as_slice()).ok();
    let dirty = run_git(["status", "--porcelain"].as_slice())
        .ok()
        .map(|s| !s.trim().is_empty());

    GitInfo {
        available: true,
        repo_root: root,
        commit: commit.map(|v| v.trim().to_string()),
        branch: branch.map(|v| v.trim().to_string()),
        dirty,
    }
}

fn run_git(args: &[&str]) -> AppResult<String> {
    let out = Command::new("git")
        .args(args)
        .current_dir(Path::new("."))
        .output()
        .map_err(|err| format!("git command failed: {err}"))?;
    if !out.status.success() {
        return Err("git command failed".to_string());
    }
    Ok(String::from_utf8_lossy(&out.stdout).to_string())
}
